File: python/39_Combination_Sum.py
#!/usr/bin/python
import sys
import os


'''
[2, 3, 6, 7] and target 7, -> [ [7], [2, 2, 3] ]
'''

class Solution(object):
    def combinationSum(self, nums, target):
    	nums = sorted(nums)
    	res, part = [], []
    	return self.dfs(nums, target, part, res)

    def dfs(self, nums, target, part, res):
    	if target == 0:
    		res.append(part)
    	n = len(nums)
    	for i in range(n):
    		if nums[i] > target:
    			break
    		self.dfs(nums[i:n], target - nums[i], part + [nums[i]], res)
        return res

def main():
    s = Solution()
    nums = [2, 3, 6, 7]
    target = 7
    print s.combinationSum(nums, target)

if __name__ == "__main__":
    main()

File: c++/39_Combination_Sum.cpp
#include <iostream>
#include <vector>
#include <map>
#include <string>
#include <algorithm>

using namespace std;

class Solution {
public:
    vector<vector<int> > combinationSum(vector<int>& num, int target) {
        vector<int> part;
        vector<vector<int> >res;
        sort(num.begin(), num.end());
        res = dfs(num, 0, target, part, res);
        return res;
    }
    // pass by reference, important, otherwise change nothing
    vector<vector<int> > dfs(vector<int> &num, int start, int target, vector<int> &part, vector<vector<int> > &res) {
    	if (target == 0)
    		res.push_back(part);
    	int n = num.size();
    	for (int i = start; i < n; i++) {
    		if (num[i] > target)
    			break;
    		part.push_back(num[i]);
    		dfs(num, i, target - num[i], part, res);
    		// difference with python, need to recover part
    		part.pop_back();
    	}
    	return res;
    }
};


int main() {
	Solution s;	
	int dat[] = {2, 3, 6, 7};
	vector<int> vec(dat, dat + sizeof(dat) / sizeof(int));
	int target = 7;

	vector<vector<int> > res = s.combinationSum(vec, target);
	for (int i = 0; i < res.size(); i++) {
		cout << "loop = " << i << " ";
		for (int j = 0; j < res[i].size(); j++)
			cout << res[i][j] << " ";
		cout << endl;
	}
}
